rfcs/20211118-support-implicit-concat/README.md

```diff
@@ -0,0 +1,139 @@
+  not fully under user's control now. But such usage is independent from memory
+  kind.
+  ![Scenario 2a](Scenario_2a.jpg)
+* Scenario 2b, identical to 2a, but also with strided input requirement for
+  certain primitives. It may happen in case of branching when output of
+  convolution should be put into extended memory and this output would be
+  further used as input in a second branch as entry point.
+  ![Branching example](branch_example.jpg)
+  ![Scenario 2b](Scenario_2b.jpg)
+
+### Implementation
+
+For scenario 2a, a simplified algorithm of actions to implement implicit concat
+performance optimization would be the following:
+
+1. Set up primitive descriptors of primitives which would contribute to concat
+   primitive.
+   a. If any of those are compute-intense primitives, querying of destination
+      memory descriptor would be required to get a proper memory format.
+   b. If all of those are memory-bound primitives, query is not required,
+      formats should already be defined and coincide or be compatible.
+2. Set up concat primitive descriptor with destination descriptors from previous
+   step. Omit concat destination to let the library deduce it.
+   a. If everything is good to go, destination would have same blocked format.
+   b. If not, concat would return `abcd`-like format.
+3. Verify that destination was created properly. This can be achieved by
+   creating a memory descriptor with concat destination dims and plain
+   `abcd`-like layout. If they match with queried, concat failed to prepare an
+   implicit setup. Consider using an explicit one.
+4. Create `submemory` descriptors from concat destination memory descriptor.
+   a. If both memory descriptors were created properly, it means that implicit
+      concat is good to go.
+   b. It may happen that a suboptimal implementation is chosen, but such
+      situations should be treated as library problems and reported.
+5. Re-create primitive descriptors of contributing primitives.
+   a. Ideally, implementations with dense and submemoried destination
+      descriptors should coincide. They may be additionally validated with
+      `impl_info_str()` call.
+6. Create new memory objects for implicit concat destination and for
+   contributing primitives. Last one should utilize a handle from the first one.
+   Pass updated memory objects for execution.
+
+Strided output only case is demonstrated at
+[this example](implicit_concat_extended_output.cpp). Both strided input and
+output case is demonstrated at another
+[example here](implicit_concat_extended_input_and_output.cpp). They have
+coverage for both scenarios depending on values provided for `IC`, `OC`, and
+`N`.
+
+### Coverage
+
+Since this feature is complex enough to maintain, support and validate,
+primitive support will be extended upon request. The initial implementation
+would support convolution and pooling with strided input and output.
```

Comment:
Do you plan to implement the support for reference implementations too?

Reply:
Actually, this support was almost there. Just needed to shift pointers. stride-wise all work was there initially when it comes to GPU. For CPU I haven't validated the status if this works as expected or not.